utils_nlp/models/transformers/abstractive_summarization_bertsum.py

```diff
@@ -0,0 +1,815 @@
+            lr_bert=learning_rate_bert,
+            warmup_steps_bert=warmup_steps_bert,
+            checkpoint=None,
+        )
+        self.optim_dec = model_builder.build_optim_dec(
+            self.model,
+            optim=optimization_method,
+            visible_gpus=None,  # ",".join([str(i) for i in range(num_gpus)]), #"0,1,2,3"
+            lr_dec=learning_rate_dec,
+            warmup_steps_dec=warmup_steps_dec,
+        )
+        # """
+
+        optimizers = [self.optim_bert, self.optim_dec]
+
+        self.amp = get_amp(fp16)
+        if self.amp:
+            self.model, optim = self.amp.initialize(
+                self.model, optimizers, opt_level=fp16_opt_level
+            )
+
+        global_step = 0
+        if checkpoint:
+            if checkpoint["optimizers"]:
+                for i in range(len(optimizers)):
+                    model_builder.load_optimizer_checkpoint(
+                        optimizers[i], checkpoint["optimizers"][i]
+                    )
+            if self.amp and "amp" in checkpoint and checkpoint["amp"]:
+                self.amp.load_state_dict(checkpoint["amp"])
+            if "global_step" in checkpoint and checkpoint["global_step"]:
+                global_step = checkpoint["global_step"] / world_size
+                print("global_step is {}".format(global_step))
+
+        self.model = parallelize_model(
+            model=self.model,
+            device=device,
+            num_gpus=num_gpus,
+            gpu_ids=gpu_ids,
+            local_rank=local_rank,
+        )
+
+        if local_rank == -1:
+            sampler = RandomSampler(train_dataset)
+        else:
+            sampler = DistributedSampler(
+                train_dataset, num_replicas=world_size, rank=rank
+            )
+
+        def collate_fn(data):
+            return self.processor.collate(
+                data, block_size=self.max_pos_length, device=device
+            )
+
+        train_dataloader = DataLoader(
+            train_dataset,
+            sampler=sampler,
+            batch_size=batch_size,
+            collate_fn=collate_fn,
+        )
+
+        # compute the max number of training steps
+        max_steps = compute_training_steps(
+            train_dataloader,
+            max_steps=max_steps,
+            gradient_accumulation_steps=gradient_accumulation_steps,
+        )
+
+        super().fine_tune(
+            train_dataloader=train_dataloader,
+            get_inputs=BertSumAbsProcessor.get_inputs,
+            device=device,
+            num_gpus=num_gpus,
+            max_steps=max_steps,
+            global_step=global_step,
+            max_grad_norm=max_grad_norm,
+            gradient_accumulation_steps=gradient_accumulation_steps,
+            verbose=verbose,
+            seed=seed,
+            report_every=report_every,
+            save_every=save_every,
+            clip_grad_norm=False,
```

Comment:
If we set clip_grad_norm to False internally, should we not expose max_grad_norm to the user? They won't be able to control it anyway. It may give the user the wrong impression that they can control it.

Reply:
remember bertsum is using it's own optimizer. clip_grad_norm is used to block the clipping in the base class. max_grad_norm can be set for the bertsum optimizer. I didn't include in the https://github.com/microsoft/nlp-recipes/blob/daden/bertsumabs/utils_nlp/models/transformers/abstractive_summarization_bertsum.py#L556. Will add it.